### PR TITLE
parse request parameters to given type, closes #311

### LINKED ISSUE #311: Parse query and form parameters into type
Create a method to parse all received parameters (forms, query, path) inside a class with the desired types, structure and checks.

The method would be something like: `fun <T: Any> Request.allParameters(type: KClass<T>): T` this method will:
1. gather all request's parameters (form, path and query ones)
2. put them in a map (taking care of the parameters that have a list of values)
3. convert the map in the supplied type
4. if conversion fails, halt the request with a 400 error

I.e. For a received request with:
* Path param: teamId: team
* Query param: userId: [ usr1, usr2, usr3 ]
* Parameter parsing type:
```kotlin
data class FilterUsers(val teamId: String, val userId: List<String>) {
  init {
    require(teamId.isNotBlank())
  }
}
```
Calling `request.allParameters(FilterUsers::class)` would return a `FilterUsers` instance with the parameters mapped.

For later revisions the error handling will be revisited to provide more details if the method fails to map the parameters to the desired type.

### PR DESCRIPTION
**Pull request by AnoojPoozhikuth**
---

For the Pull Request to be accepted please check:

- [x] The PR has a meaningful title.
- [x] If the PR refers to an issue, it should be referenced with the Github format (*#ID*).
- [x] The PR is done to the `develop` branch.
- [x] The code pass all PR checks.
- [x] All public members are documented using [Dokka](https://github.com/Kotlin/dokka).
- [x] The code follow the coding conventions stated at the [contributing.md](/contributing.md) file.

### CODE REVIEW

File: port_http_client/src/main/kotlin/Request.kt [hunk 1]
```diff
@@ -37,3 +40,35 @@ data class Request(
...
+    }
+}
+
+/**
+ * inline function used to aggreate all request parameters and create a map object out of it
+ * @param request : an object of class com.hexagonkt.http.client.Request
+ * @return a map containing all request paremeters
+ */
+private inline fun generateRequestMap(request: Request): Map<String, Any> {
+    val requestMap: MutableMap<String, Any> = hashMapOf()
```
Comment (jaguililla) [inline port_http_client/src/main/kotlin/Request.kt:68]: Nice, I'm only missing the addition of query parameters :)

File: port_http_server/src/main/kotlin/Request.kt [hunk 2]
```diff
@@ -87,3 +89,46 @@ abstract class Request {
...
+ * This function aggregates path parameters, form parameters and query parameters into a map and convert it into given class using object mapper
+ * Usage : request.parseAllParameters(MyCustomDataClass::class)
+ *
+ * @param type is the KotlinClass of type T (where T can be any class eg:MyCustomDataClass)
+ * @return an object of type T  (eg: MyCustomDataClass())
+ */
+fun <T : Any> Request.parseAllParameters(type: KClass<T>): T? {
+    val requestMap = generateRequestMap(this)
+    return try {
+        jacksonObjectMapper().convertValue(requestMap, type.javaObjectType)
```
Comment (jaguililla) [inline port_http_server/src/main/kotlin/Request.kt:103]: You can use the `hexagon_core` helper method `parse` in the `serialization` package: https://hexagonkt.com/hexagon_core/#serialization

**Comment by jaguililla** [comment]
Good job! however, I'm missing the tests ;)

File: port_http_server/src/main/kotlin/Request.kt (hunk 2 above)
Comment (AnoojPoozhikuth) [reply port_http_server/src/main/kotlin/Request.kt:103]: Ok @jaguililla  , gonna replace the  _jacksonObjectMapper().convertValue(requestMap, type.javaObjectType)_ with the existing method as you mentioned 

**requestMap.convertToObject<T>(type)**

File: port_http_client/src/main/kotlin/Request.kt (hunk 1 above)
Comment (AnoojPoozhikuth) [reply port_http_client/src/main/kotlin/Request.kt]: We can ignore this, I moved it to the right place , which has query parameters 
com.hexagonkt.http.server.Request

**Comment by AnoojPoozhikuth** [comment]
> Good job! however, I'm missing the tests ;)
Added :)

Please review now

**Comment by AnoojPoozhikuth** [comment]
Added one more test where multiple parameters are part of the same map , please review now @jaguililla

**Comment by jaguililla** [comment]
Merged... Thank you very much!

**Pull request merged**